File: packages/clay_core/test/src/features/config/clay_compatibility_test.dart
import 'package:clay_core/clay.dart';
import 'package:clay_core/config.dart';
import 'package:pub_semver/pub_semver.dart';
import 'package:test/test.dart';

void main() {
  group('isClayConfigCompatibleWithClay', () {
    test('returns true when environment.clay is any', () {
      expect(isClayConfigCompatibleWithClay(ClayConfig()), isTrue);
    });

    test('returns true when constraint allows the current version', () {
      final config = ClayConfig(
        environment: ClayEnvironment(
          clay: VersionConstraint.parse('^0.0.1-dev.1'),
        ),
      );

      expect(isClayConfigCompatibleWithClay(config), isTrue);
    });

    test('returns false when constraint excludes the current version', () {
      final config = ClayConfig(
        environment: ClayEnvironment(
          clay: VersionConstraint.parse('^0.2.0'),
        ),
      );

      expect(isClayConfigCompatibleWithClay(config), isFalse);
    });
  });

  group('assertClayCompatible', () {
    test('does not throw when environment.clay is any', () {
      expect(() => assertClayCompatible(ClayConfig()), returnsNormally);
    });

    test('does not throw when constraint allows the current version', () {
      final config = ClayConfig(
        environment: ClayEnvironment(
          clay: VersionConstraint.parse('^0.0.1-dev.1'),
        ),
      );

      expect(() => assertClayCompatible(config), returnsNormally);
    });

    test('throws ClayIncompatibleException with actionable message', () {
      final config = ClayConfig(
        environment: ClayEnvironment(
          clay: VersionConstraint.parse('^0.2.0'),
        ),
      );

      expect(
        () => assertClayCompatible(config),
        throwsA(
          isA<ClayIncompatibleException>()
              .having(
                (error) => error.currentVersion,
                'currentVersion',
                packageVersion,
              )
              .having(
                (error) => error.requiredConstraint,
                'requiredConstraint',
                '^0.2.0',
              )
              .having(
                (error) => error.toString(),
                'message',
                'The current clay version is $packageVersion.\n'
                'This project requires clay version ^0.2.0.',
              ),
        ),
      );
    });
  });

  group('environment.clay parsing', () {
    test('rejects invalid constraint strings before compatibility checks', () {
      expect(
        () => ClayEnvironment.fromMap(const {'clay': 'not-a-version'}),
        throwsA(
          predicate<Object>(
            (error) => error.toString().contains(
              'environment.clay must be a valid semver constraint',
            ),
          ),
        ),
      );
    });
  });
}

File: extensions/clay/test/clayCompatibility.test.mjs
import assert from 'node:assert/strict';
import { createRequire } from 'node:module';
import { test } from 'node:test';

const require = createRequire(import.meta.url);

const {
  DEFAULT_CLAY_CONSTRAINT,
  assertClayConfigCompatibleWithCli,
  formatClayIncompatibleMessage,
  isClayConfigCompatibleWithCliVersion,
  validateClayVersionConstraint,
} = require('./out/clayCompatibility.cjs');
const { parseClayConfig } = require('./out/clayConfig.cjs');

function configWithClayConstraint(constraint) {
  return parseClayConfig(`environment:\n  clay: ${constraint}\n`);
}

test('validateClayVersionConstraint accepts any and valid semver ranges', () => {
  assert.doesNotThrow(() => validateClayVersionConstraint(DEFAULT_CLAY_CONSTRAINT));
  assert.doesNotThrow(() => validateClayVersionConstraint('^0.0.1-dev.1'));
});

test('validateClayVersionConstraint rejects empty and invalid constraints', () => {
  assert.throws(
    () => validateClayVersionConstraint(''),
    /environment\.clay must not be empty/,
  );
  assert.throws(
    () => validateClayVersionConstraint('not-a-version'),
    /environment\.clay must be a valid semver constraint/,
  );
});

test('isClayConfigCompatibleWithCliVersion treats any as satisfied', () => {
  const config = parseClayConfig('');
  assert.equal(isClayConfigCompatibleWithCliVersion(config, '0.0.1-dev.1'), true);
});

test('isClayConfigCompatibleWithCliVersion checks satisfied and unsatisfied ranges', () => {
  const compatible = configWithClayConstraint('^0.0.1-dev.1');
  const incompatible = configWithClayConstraint('^99.0.0');

  assert.equal(isClayConfigCompatibleWithCliVersion(compatible, '0.0.1-dev.1'), true);
  assert.equal(isClayConfigCompatibleWithCliVersion(incompatible, '0.0.1-dev.1'), false);
});

test('assertClayConfigCompatibleWithCli throws an actionable message', () => {
  const config = configWithClayConstraint('^99.0.0');

  assert.throws(
    () => assertClayConfigCompatibleWithCli(config, '0.0.1-dev.1'),
    (error) => {
      assert.equal(
        error.message,
        formatClayIncompatibleMessage('0.0.1-dev.1', '^99.0.0'),
      );
      return true;
    },
  );
});

test('formatClayIncompatibleMessage matches the Clay CLI wording', () => {
  assert.equal(
    formatClayIncompatibleMessage('0.0.1-dev.1', '^99.0.0'),
    'The current clay version is 0.0.1-dev.1.\nThis project requires clay version ^99.0.0.',
  );
});
